### PR TITLE
Add global exception handler

### PR DESCRIPTION
**Pull request by EduarfF**

### CODE REVIEW

File: src/main/java/org/example/springbootintro/dto/book/BookSearchParametersDto.java [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package org.example.springbootintro.dto.book;
+
+public record BookSearchParametersDto(String[] title, String[] author) {
+}
```
Comment (TonyH277) [inline src/main/java/org/example/springbootintro/dto/book/BookSearchParametersDto.java]: This is not part of the global exception handler, you should create pull request from this brach to add_criteria_query_book not to main.
Check this out https://mate-academy.github.io/jv-program-fulltime/07_spring_boot/git-strategy

File: src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.example.springbootintro.dto.book.BookSearchParametersDto;
+import org.example.springbootintro.model.Book;
+import org.example.springbootintro.repository.SpecificationBuilder;
+import org.example.springbootintro.repository.SpecificationProviderManager;
+import org.example.springbootintro.repository.book.spec.AuthorSpecificationProvider;
+import org.example.springbootintro.repository.book.spec.TitleSpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@AllArgsConstructor
```
Comment (TonyH277) [inline src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java]:
```suggestion
@RequiredArgsConstructor
```
as you wanna have only required properties injected

**Review by TonyH277 — CHANGES_REQUESTED** [review]
I think all good with your implementation, but you shoud create pull request to add_criteria_query_book not to main brach

File: pom.xml [hunk 3]
```diff
@@ -60,6 +60,10 @@
 			<artifactId>liquibase-core</artifactId>
 			<version>${liquibase.version}</version>
 		</dependency>
+		<dependency>
+			<groupId>org.hibernate.validator</groupId>
+			<artifactId>hibernate-validator</artifactId>
+		</dependency>
```
Comment (andrii-hoienko) [inline pom.xml]: Why this dependency is required?

File: src/main/java/org/example/springbootintro/controller/BookController.java [hunk 4]
```diff
@@ -21,10 +26,16 @@
 @RequiredArgsConstructor
 public class BookController {
     private final BookService bookService;
+    private final BookSearchParametersValidator bookSearchParametersValidator;
 
-    @PostMapping
+    @InitBinder("bookSearchParametersDto")
+    public void initBinder(WebDataBinder binder) {
+        binder.addValidators(bookSearchParametersValidator);
+    }
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/controller/BookController.java:34]: Why this is required? Why configuration is performed in controller?

File: src/main/java/org/example/springbootintro/controller/BookController.java [hunk 5]
```diff
@@ -51,4 +62,10 @@ public void deleteById(@PathVariable Long id) {
     public BookDto updateById(@PathVariable Long id, @RequestBody CreateBookRequestDto bookDto) {
         return bookService.updateById(id, bookDto);
     }
+
+    @GetMapping("/search")
+    @ResponseStatus(HttpStatus.OK)
+    public List<BookDto> search(@Valid BookSearchParametersDto bookSearchParametersDto) {
+        return bookService.search(bookSearchParametersDto);
+    }
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/controller/BookController.java]: This changes is not related to task

File: src/main/java/org/example/springbootintro/controller/BookController.java [hunk 6]
```diff
@@ -51,4 +62,10 @@ public void deleteById(@PathVariable Long id) {
     public BookDto updateById(@PathVariable Long id, @RequestBody CreateBookRequestDto bookDto) {
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/controller/BookController.java]:
```suggestion
    public BookDto updateById(@PathVariable Long id, @RequestBody @Valid CreateBookRequestDto bookDto) {
```

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Please make PR with all changes related to task

File: pom.xml (hunk 3 above)
Comment (EduarfF) [reply pom.xml]: To use:
import jakarta.validation.Valid;

@Valid

File: src/main/java/org/example/springbootintro/controller/BookController.java (hunk 5 above)
Comment (EduarfF) [reply src/main/java/org/example/springbootintro/controller/BookController.java]: Add meaningful validation for existing DTO classes
I add my own validator. To search by title or author or both, and see error message about empty field and so on.

File: src/main/java/org/example/springbootintro/controller/BookController.java (hunk 4 above)
Comment (EduarfF) [reply src/main/java/org/example/springbootintro/controller/BookController.java:34]: I need this to correct working of my own validator: BookSearchParametersValidator

File: src/main/java/org/example/springbootintro/controller/BookController.java [hunk 7]
```diff
@@ -21,10 +26,16 @@
...
     private final BookService bookService;
+    private final BookSearchParametersValidator bookSearchParametersValidator;
 
-    @PostMapping
+    @InitBinder("bookSearchParametersDto")
+    public void initBinder(WebDataBinder binder) {
+        binder.addValidators(bookSearchParametersValidator);
+    }
+
+    @PostMapping("/add")
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/controller/BookController.java]: `/add` is not very standard for RESTful applications.
It will be better just having POST `/books`

File: src/main/java/org/example/springbootintro/controller/BookController.java [hunk 8]
```diff
@@ -48,7 +59,15 @@ public void deleteById(@PathVariable Long id) {
...
+                              @RequestBody @Valid CreateBookRequestDto bookDto
+    ) {
         return bookService.updateById(id, bookDto);
     }
+
+    @GetMapping("/search")
+    @ResponseStatus(HttpStatus.OK)
+    public List<BookDto> search(@Valid BookSearchParametersDto bookSearchParametersDto) {
+        return bookService.search(bookSearchParametersDto);
+    }
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/controller/BookController.java]: this PR is about global exception handler. It should not have any other functionality in it. 
Please merge main into this branch, I think it should fix it, since `feat 4` was merged

File: src/main/java/org/example/springbootintro/dto/book/validator/BookSearchParametersValidator.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+@Component
+public class BookSearchParametersValidator implements Validator {
+    @Override
+    public boolean supports(Class<?> clazz) {
+        return BookSearchParametersDto.class.equals(clazz);
+    }
+
+    @Override
+    public void validate(Object target, Errors errors) {
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/dto/book/validator/BookSearchParametersValidator.java:16]: I think you are inventing a wheel
Why don't you just add some annotations from `javax.validation` to `BookSearchParametersDto`?

File: src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java (hunk 2 above)
Comment (yevheniisynytsia) [reply src/main/java/org/example/springbootintro/repository/book/BookSpecificationBuilder.java]: https://github.com/EduarfF/book-store/pull/6/files#r1585980458
not fixed

File: src/main/resources/application.properties [hunk 10]
```diff
@@ -1,7 +1,7 @@
 spring.application.name=spring-boot-intro
 spring.datasource.url=jdbc:mysql://localhost:3306/bookstore?serverTimezone=UTC
-spring.datasource.username=
-spring.datasource.password=
+spring.datasource.username=root
+spring.datasource.password=GwFewBr2ty20B5SF
```
Comment (yevheniisynytsia) [inline src/main/resources/application.properties]: you can use environment variables to not to push credentials.
You may take a look at the first QnA recording from 29.04, where we set it up

File: pom.xml [hunk 11]
```diff
@@ -60,6 +60,10 @@
 			<artifactId>liquibase-core</artifactId>
 			<version>${liquibase.version}</version>
 		</dependency>
+		<dependency>
+			<groupId>org.hibernate.validator</groupId>
+			<artifactId>hibernate-validator</artifactId>
```
Comment (yevheniisynytsia) [inline pom.xml]: does the validation work only with this dependency?
I ask because there may be issues with different versions of this lib. The solution to it was using `springboot-starter-validation`. But if it works for you, let it be so

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]
Well done! Please try to prepare the PR so that it contains only changes related to the task it was created for. I also showed how to merge main on the first QnA on 29.04

File: src/main/java/org/example/springbootintro/dto/book/validator/BookSearchParametersValidator.java (hunk 9 above)
Comment (EduarfF) [reply src/main/java/org/example/springbootintro/dto/book/validator/BookSearchParametersValidator.java:16]: I want to use or only title or author or both. If I will use annotations @NotEmpty or something else, it will fail. Bohdan said that I need to create custom.

File: src/main/java/org/example/springbootintro/controller/BookController.java (hunk 8 above)
Comment (EduarfF) [reply src/main/java/org/example/springbootintro/controller/BookController.java]: But it is about:
1) Add meaningful validation for existing DTO classes
So I create my own validator for my search request
Maybe I don't understand something, because as for me it's belongs to this task too

File: src/main/java/org/example/springbootintro/dto/book/validator/BookSearchParametersValidator.java (hunk 9 above)
Comment (yevheniisynytsia) [reply src/main/java/org/example/springbootintro/dto/book/validator/BookSearchParametersValidator.java:16]: Ok, let's leave it

File: src/main/java/org/example/springbootintro/controller/BookController.java (hunk 8 above)
Comment (yevheniisynytsia) [reply src/main/java/org/example/springbootintro/controller/BookController.java]: yes, you have to update some code related to search in this PR.
But if you open changed files, you see that this branch wants to add files that are already in main, they were merged in this PR https://github.com/EduarfF/book-store/pull/5/files#diff-7c08aaee8faaab30e0062c476037503493f9442f327f45d99df48fbde0dbb1f4

New PRs shouldn't have files that were already merged. I think you have to merge main into your branch and this will fix the PR.
Come to QnA today to resolve this issue

Comment (EduarfF) [reply src/main/java/org/example/springbootintro/controller/BookController.java]: Tried to do rebase with Vadym
